src/wp-includes/speculative-loading.php

```diff
@@ -0,0 +1,245 @@
+<?php
+/**
+ * Speculative loading functions.
+ *
+ * @package WordPress
+ * @subpackage Speculative Loading
+ * @since 6.8.0
+ */
+
+/**
+ * Returns the speculation rules configuration.
+ *
+ * @since 6.8.0
+ *
+ * @return array<string, string>|false Associative array with 'mode' and 'eagerness' keys, or false if speculative
+ *                                     loading is disabled.
+ */
+function wp_get_speculation_rules_configuration() {
+	$config = array(
+		'mode'      => 'auto',
+		'eagerness' => 'auto',
+	);
+
+	/**
+	 * Filters the way that speculation rules are configured.
+	 *
+	 * The Speculation Rules API is a web API that allows to automatically prefetch or prerender certain URLs on the
+	 * page, which can lead to near-instant page load times. This is also referred to as speculative loading.
+	 *
+	 * There are two aspects to the configuration:
+	 * * The "mode" (whether to "prefetch" or "prerender" URLs).
+	 * * The "eagerness" (whether to speculatively load URLs in an "eager", "moderate", or "conservative" way).
+	 *
+	 * By default, the speculation rules configuration is decided by WordPress Core ("auto"). This filter can be used
+	 * to force a certain configuration, which could for instance load URLs more or less eagerly.
+	 *
+	 * @since 6.8.0
+	 * @see https://developer.chrome.com/docs/web-platform/prerender-pages
+	 *
+	 * @param array<string, string>|false $config Associative array with 'mode' and 'eagerness' keys. The default value
+	 *                                            for both of them is 'auto'. Other possible values for 'mode' are
+	 *                                            'prefetch' and 'prerender'. Other possible values for 'eagerness' are
+	 *                                            'eager', 'moderate', and 'conservative'. Alternatively, you may
+	 *                                            return `false` to disable speculative loading entirely.
+	 */
+	$config = apply_filters( 'wp_speculation_rules_configuration', $config );
+
+	// Allow the value `false` to indicate that speculative loading is disabled.
+	if ( false === $config ) {
+		return false;
+	}
+
+	// Sanitize the configuration and replace 'auto' with current defaults.
+	$default_mode      = 'prefetch';
+	$default_eagerness = 'conservative';
+	if ( ! is_array( $config ) ) {
+		return array(
+			'mode'      => $default_mode,
+			'eagerness' => $default_eagerness,
+		);
+	}
```

Comment:
No need for this if the above check for `! is_array( $config )` short circuits:

```suggestion
```

Reply:
See my reply above, I think we need to keep those two as separate checks.